allennlp/data/data_loaders/pytorch_data_loader.py

```diff
@@ -0,0 +1,224 @@
+from typing import List, Iterator
+
+from torch.utils import data
+
+from allennlp.common.lazy import Lazy
+from allennlp.common.tqdm import Tqdm
+from allennlp.data.instance import Instance
+from allennlp.data.dataset_readers import DatasetReader, WorkerInfo
+from allennlp.data.samplers import PyTorchSampler, PyTorchBatchSampler
+from allennlp.data.vocabulary import Vocabulary
+from allennlp.data.data_loaders.data_loader import DataLoader, allennlp_collate
+
+
+class AllennlpDataset(data.Dataset):
+    def __init__(self, instances: List[Instance], vocab: Vocabulary = None):
+        self.instances = instances
+        self.vocab = vocab
+
+    def __getitem__(self, idx) -> Instance:
+        if self.vocab is not None:
+            self.instances[idx].index_fields(self.vocab)
+        return self.instances[idx]
+
+    def __len__(self):
+        return len(self.instances)
+
+    def __iter__(self) -> Iterator[Instance]:
+        """
+        Even though it's not necessary to implement this because Python can infer
+        this method from `__len__` and `__getitem__`, this helps with type-checking
+        since `AllennlpDataset` can be considered an `Iterable[Instance]`.
+        """
+        yield from self.instances
```

Comment:
There is a semantic difference though. If you pretend to be an `Iterable` through `__getitem__`, it runs `index_fields()`. Here, it does not.

Reply:
Good catch!